sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnnecessaryImportsCheck.kt

```diff
@@ -0,0 +1,174 @@
+private const val MESSAGE_REDUNDANT = "Remove redundant import."
+
+@Rule(key = "S1128")
+class UnnecessaryImportsCheck : AbstractCheck() {
+
+    private fun Sequence<PsiElement>.flattenNodes(): Sequence<PsiElement> =
+        this + this.flatMap { it.children.asSequence().flattenNodes() }
+
+    override fun visitKtFile(file: KtFile, context: KotlinFileContext) {
+
+        val (references, arrayAccesses, kDocLinks) = collectData(file)
+        val bindingContext = context.bindingContext
+        val groupedReferences = references.groupBy { reference ->
+            reference.importableSimpleName()
+        }
+
+        val arrayAccessesImportsFilter by lazy {
+            getArrayAccessImportsFilter(arrayAccesses, bindingContext)
+        }
+
+        file.importDirectives.filter { imp ->
+            // 1. Filter out & report all imports that import from kotlin.* or the same package as our file
+            if (imp.isImportedImplicitlyAlready(file.packageDirective?.name)) {
+                imp.importedReference?.let { context.reportIssue(it, MESSAGE_REDUNDANT) }
+                false
+            } else true
+        }.groupBy {
+            it.importedName?.asString()
+        }.filterKeys { simpleName ->
+            // 2. Discard all imports that could be relevant for KDocs. This is done a bit fuzzy.
+            simpleName != null && simpleName !in kDocLinks
+        }.flatMap { (simpleName, importsWithSameName) ->
+            // 3. Discard all imports that are used in references throughout the code. With binding context if possible,
+            // otherwise we over-estimate
+            groupedReferences[simpleName]?.let { relevantReferences ->
+                var relevantImports = importsWithSameName
+                for (ref in relevantReferences) {
+                    val refName = bindingContext.get(BindingContext.REFERENCE_TARGET, ref)?.getImportableDescriptor()?.fqNameOrNull()
+                        ?: return@flatMap emptyList() // Discard all: over-estimate, resulting in less FPs and more FNs without binding ctx
+                    relevantImports = relevantImports.filter { it.importedFqName != refName }
+                    if (relevantImports.isEmpty()) break
+                }
+                relevantImports
+            } ?: importsWithSameName
+        }.filter {
+            // 4. Filter 'get' and 'set' imports
+            arrayAccessesImportsFilter(it)
+        }.forEach { importDirective ->
+            // We could not find any usages for anything remaining at this point. Hence, report!
+            importDirective.importedReference?.let{ context.reportIssue(it, MESSAGE_UNUSED) }
+        }
+    }
+
+    private fun collectData(file: KtFile) =
+        file.children.asSequence().filter {
+            it !is KtPackageDirective && it !is KtImportList
+        }.let { relevantTopLevelChildren ->
+            DataCollector(file).apply {
+                relevantTopLevelChildren.flattenNodes()
```

Comment:
Do we need really to flattenNodes? It looks like you can use `KtTreeVisitorVoid` and visit children of elements